engine/access/rest/middleware/metrics.go

```diff
@@ -0,0 +1,139 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/slok/go-http-metrics/metrics"
+	metricsProm "github.com/slok/go-http-metrics/metrics/prometheus"
+	"github.com/slok/go-http-metrics/middleware"
+	"github.com/slok/go-http-metrics/middleware/std"
+
+	"github.com/gorilla/mux"
+)
+
+// Example recorder taken from:
+// https://github.com/slok/go-http-metrics/blob/master/metrics/prometheus/prometheus.go
+type CustomRecorder interface {
+	metrics.Recorder
+	AddTotalRequests(ctx context.Context, service string, id string)
+}
+
+type recorder struct {
+	httpRequestDurHistogram   *prometheus.HistogramVec
+	httpResponseSizeHistogram *prometheus.HistogramVec
+	httpRequestsInflight      *prometheus.GaugeVec
+	httpRequestsTotal         *prometheus.GaugeVec
+}
+
+// NewRecorder returns a new metrics recorder that implements the recorder
+// using Prometheus as the backend.
+func NewRecorder(cfg metricsProm.Config) CustomRecorder {
+	if len(cfg.DurationBuckets) == 0 {
+		cfg.DurationBuckets = prometheus.DefBuckets
+	}
+
+	if len(cfg.SizeBuckets) == 0 {
+		cfg.SizeBuckets = prometheus.ExponentialBuckets(100, 10, 8)
+	}
+
+	if cfg.Registry == nil {
+		cfg.Registry = prometheus.DefaultRegisterer
+	}
+
+	if cfg.HandlerIDLabel == "" {
+		cfg.HandlerIDLabel = "handler"
+	}
+
+	if cfg.StatusCodeLabel == "" {
+		cfg.StatusCodeLabel = "code"
+	}
+
+	if cfg.MethodLabel == "" {
+		cfg.MethodLabel = "method"
+	}
+
+	if cfg.ServiceLabel == "" {
+		cfg.ServiceLabel = "service"
+	}
+
+	r := &recorder{
+		httpRequestDurHistogram: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: cfg.Prefix,
+			Subsystem: "http",
+			Name:      "request_duration_seconds",
+			Help:      "The latency of the HTTP requests.",
+			Buckets:   cfg.DurationBuckets,
+		}, []string{cfg.ServiceLabel, cfg.HandlerIDLabel, cfg.MethodLabel, cfg.StatusCodeLabel}),
```

Comment:
do handler and method map 1:1 or are there cases where they are different?

Reply:
Method in this case is the type of HTTP method (like `GET` or `POST`) while Handler is the actual HTTP path of the request (like `/v1/blocks` for example). And both are always present in the HTTP request, if that answers your question, let me know 👀